File: test_yt_methods.py
import re
import xml.etree.ElementTree as ET
import requests
import urllib3

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

PLAYLIST_ID = "PLSqr4-dP7wJDARLGADNTbBn3H-2lgSoF4"

print("=== Method 1: RSS Feed ===")
rss_url = f"https://www.youtube.com/feeds/videos.xml?playlist_id={PLAYLIST_ID}"
try:
    resp = requests.get(rss_url, timeout=15, verify=False)
    if resp.status_code == 200:
        root = ET.fromstring(resp.text)
        entries = root.findall("{http://www.w3.org/2005/Atom}entry")
        print(f"RSS entries found: {len(entries)}")
        for entry in entries[:5]:
            title = entry.find("{http://www.w3.org/2005/Atom}title").text
            yt_id = entry.find("{http://www.youtube.com/xml/schemas/2015}videoId").text
            print(f"  [{yt_id}] {title}")
    else:
        print(f"RSS Status code: {resp.status_code}")
except Exception as e:
    print(f"RSS error: {e}")

print("\n=== Method 2: Mobile User-Agent ===")
mobile_headers = {
    "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 16_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Mobile/15E148 Safari/604.1",
    "Accept-Language": "ko-KR,ko;q=0.9",
}
try:
    url = f"https://www.youtube.com/playlist?list={PLAYLIST_ID}"
    resp = requests.get(url, headers=mobile_headers, timeout=15, verify=False)
    vids = set(re.findall(r'/watch\?v=([a-zA-Z0-9_-]{11})', resp.text))
    print(f"Mobile watch IDs found: {len(vids)}")
except Exception as e:
    print(f"Mobile error: {e}")

print("\n=== Method 3: Desktop Page ytInitialData / HTML ===")
desktop_headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    "Accept-Language": "ko-KR,ko;q=0.9",
    "Cookie": "CONSENT=YES+1",
}
try:
    url = f"https://www.youtube.com/playlist?list={PLAYLIST_ID}"
    resp = requests.get(url, headers=desktop_headers, timeout=15, verify=False)
    html = resp.text
    print(f"Desktop HTML length: {len(html)}")
    
    # Check for ytInitialData
    m = re.search(r'ytInitialData\s*=\s*({.*?});</script>', html)
    if not m:
        m = re.search(r'window\["ytInitialData"\]\s*=\s*({.*?});', html)
    
    if m:
        print("Found ytInitialData pattern!")
        # count playlistVideoRenderer occurrences in raw text
        pvrs = re.findall(r'"playlistVideoRenderer":\s*({.*?}),"playlistVideoRenderer"', html)
        print(f"Raw playlistVideoRenderer count: {len(pvrs)}")
    else:
        print("ytInitialData pattern NOT matched in desktop html.")
except Exception as e:
    print(f"Desktop error: {e}")

print("\n=== Method 4: yt-dlp check ===")
try:
    import yt_dlp
    ydl_opts = {'extract_flat': True, 'quiet': True}
    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        info = ydl.extract_info(f"https://www.youtube.com/playlist?list={PLAYLIST_ID}", download=False)
        entries = info.get('entries', [])
        print(f"yt-dlp found {len(entries)} entries!")
        for e in entries[:5]:
            print(f"  [{e.get('id')}] {e.get('title')}")
except ImportError:
    print("yt-dlp is not installed in current environment.")
except Exception as e:
    print(f"yt-dlp error: {e}")

File: test_yt_methods.js
const https = require('https');

const playlistId = 'PLSqr4-dP7wJDARLGADNTbBn3H-2lgSoF4';

// Test 1: RSS Feed
const rssUrl = `https://www.youtube.com/feeds/videos.xml?playlist_id=${playlistId}`;
https.get(rssUrl, (res) => {
  let data = '';
  res.on('data', chunk => data += chunk);
  res.on('end', () => {
    console.log(`[RSS] Status: ${res.statusCode}, Length: ${data.length}`);
    const matches = [...data.matchAll(/<entry>[\s\S]*?<yt:videoId>(.*?)<\/yt:videoId>[\s\S]*?<title>(.*?)<\/title>[\s\S]*?<\/entry>/g)];
    console.log(`[RSS] Entries count: ${matches.length}`);
    matches.forEach(m => {
      console.log(`  [RSS Item] ${m[1]} -> ${m[2]}`);
    });

    // Test 2: HTML Page ytInitialData
    testHtmlPage();
  });
}).on('error', err => console.error('[RSS] Error:', err.message));

function testHtmlPage() {
  const options = {
    hostname: 'www.youtube.com',
    path: `/playlist?list=${playlistId}`,
    headers: {
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
      'Accept-Language': 'ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7',
      'Cookie': 'CONSENT=YES+cb.20210328-17-p0.ko+FX+430'
    }
  };

  https.get(options, (res) => {
    let html = '';
    res.on('data', chunk => html += chunk);
    res.on('end', () => {
      console.log(`[HTML] Status: ${res.statusCode}, Length: ${html.length}`);
      const m = html.match(/var ytInitialData\s*=\s*({[\s\S]*?});<\/script>/);
      if (m) {
        console.log(`[HTML] Found ytInitialData! Length of JSON string: ${m[1].length}`);
        try {
          const json = JSON.parse(m[1]);
          const videos = [];
          function walk(node) {
            if (!node) return;
            if (typeof node === 'object') {
              if (node.playlistVideoRenderer) {
                const pvr = node.playlistVideoRenderer;
                const vid = pvr.videoId;
                const title = (pvr.title && pvr.title.runs) ? pvr.title.runs.map(r => r.text).join('') : '';
                if (vid && title) videos.push({ vid, title });
              }
              for (const k of Object.keys(node)) walk(node[k]);
            }
          }
          walk(json);
          console.log(`[HTML] Parsed ${videos.length} videos from ytInitialData:`);
          videos.forEach(v => console.log(`  [HTML Item] ${v.vid} -> ${v.title}`));
        } catch (e) {
          console.error('[HTML] JSON parse error:', e.message);
        }
      } else {
        console.log('[HTML] var ytInitialData regex NOT matched');
      }
    });
  }).on('error', err => console.error('[HTML] Error:', err.message));
}
